src/lib/sql/sql_pipeline.cpp

```diff
@@ -1,68 +1,105 @@
-  DebugAssert(_transaction_context != nullptr, "Cannot pass nullptr as explicit transaction context.");
-}
+SQLPipeline::SQLPipeline(const std::string& sql, bool use_mvcc) : SQLPipeline(sql, nullptr, use_mvcc) {}
 
-SQLPipeline::SQLPipeline(const std::string& sql, const bool use_mvcc)
-    : _sql_string(sql), _use_mvcc(use_mvcc), _auto_commit(_use_mvcc) {
-  if (_use_mvcc) {
-    // We want to use MVCC but didn't pass an explicit context
-    _transaction_context = TransactionManager::get().new_transaction_context();
-  } else {
-    // We don't want to use MVCC
-    _transaction_context = nullptr;
-  }
+SQLPipeline::SQLPipeline(const std::string& sql, std::shared_ptr<opossum::TransactionContext> transaction_context)
+    : SQLPipeline(sql, std::move(transaction_context), true) {
+  DebugAssert(_sql_pipeline_statements.front()->transaction_context() != nullptr,
+              "Cannot pass nullptr as explicit transaction context.");
 }
 
-const hsql::SQLParserResult& SQLPipeline::get_parsed_sql() {
-  if (_parsed_sql) {
-    // Returned cached result
-    return *(_parsed_sql.get());
-  }
-
-  auto parse_result = std::make_unique<hsql::SQLParserResult>();
-
-  const auto started = std::chrono::high_resolution_clock::now();
+// Private constructor
+SQLPipeline::SQLPipeline(const std::string& sql, std::shared_ptr<TransactionContext> transaction_context, bool use_mvcc)
+    : _transaction_context(std::move(transaction_context)) {
+  hsql::SQLParserResult parse_result;
   try {
-    hsql::SQLParser::parse(_sql_string, parse_result.get());
+    hsql::SQLParser::parse(sql, &parse_result);
   } catch (const std::exception& exception) {
     throw std::runtime_error("Error while parsing SQL query:\n  " + std::string(exception.what()));
   }
 
-  if (!parse_result->isValid()) {
-    throw std::logic_error("SQL query not valid.");
+  if (!parse_result.isValid()) {
+    throw std::runtime_error(SQLPipelineStatement::create_parse_error_message(sql, parse_result));
+  }
+
+  DebugAssert(parse_result.size() > 0, "Cannot create empty SQLPipeline.");
+  _sql_pipeline_statements.reserve(parse_result.size());
+
+  bool seen_altering_statement = false;
+
+  for (auto* statement : parse_result.releaseStatements()) {
+    switch (statement->type()) {
+      // Check if statement alters table in any way that would possibly not be seen in a later statement.
+      case hsql::StatementType::kStmtImport:
+      case hsql::StatementType::kStmtCreate:
+      case hsql::StatementType::kStmtDrop:
+      case hsql::StatementType::kStmtAlter:
+      case hsql::StatementType::kStmtRename:
+        seen_altering_statement = true;
+      default: { /* do nothing */
+      }
+    }
+
+    auto parsed_statement = std::make_shared<hsql::SQLParserResult>(statement);
+    // We know this is always true because the check above would fail otherwise
+    parsed_statement->setIsValid(true);
```

Comment:
Because it is false by default and having a valid statement `.is_valid()` return false is contradictiing.

Reply:
Ah, I misread the comment. "always true" was "always" for me, i.e., it was true before and it is true after the method call.

I think this might be one of the few cases where it would be better to leave the comment out.